### ISSUE TITLE
["Request"] New syntax for partial application

### ISSUE THREAD

**Issue by ferranpujolcamins** [bow-swift/bow#652]
## Description

It's annoying when I can't write a function point-free style just because I cannot partially apply the second argument of a 2-ary function (or more generally, the ith parameter of an n-ary). For example, we need the `lift` function in `Functor` just because we can't partially apply the function in `map` in a convenient and easy way.

I propose a new syntax for partial application that extends the current one to allow partial application of any parameters in a function.

## Sample usage

I propose to add a new operator `<|`, that takes a function as first argument and a tuple with values and placeholders as second argument. The placeholders (a placeholder is two underscores) help determine what parameter a value is being applied to.

For partial application of the first parameter, we can also keep the current simplified syntax.

```swift
let f: (String, Int) -> Int = { s, i in i }

// Partially apply first argument
let p1CurrentSyntax /* : (Int) -> Int */    = f <| "a"
let p1NewSyntax     /* : (Int) -> Int */    = f <| ("a", __)

// Partially apply second argument
let p2              /* : (String) -> Int */ = f <| (__, 1)
```

Note how I've written the resulting function type in comments, meaning that type inference works well with the syntax.

I chose to reverse the direction and parameter position of the current `|>` because now the position of the function and the tuple of arguments resembles the order in which you normally write function application.

## Potential implementation

The trick is to define a special `PlaceHolder` type just so we can write `.__` on our tuples. 

If we also define `__` as a global value we can avoid the leading dot. This is not needed (type inference also works if we write `.__`). It makes the syntax nicer, but it pollutes the global namespace. What do you think?

```swift
public enum PlaceHolder {
    case __
}

public let __ = PlaceHolder.__

infix operator <| : AdditionPrecedence

public func <|<A, B, C>(_ f: (A, B) -> C, _ values: (PlaceHolder, B)) -> (A) -> C {
    { a in f(a, values.1)}
}

public func <|<A, B, C>(_ f: (A, B) -> C, _ values: (A, PlaceHolder)) -> (B) -> C {
    { b in f(values.0, b)}
}

// ... implementations for higher arities
```

## Modules

Bow

## Breaking changes

None, unless we remove `|>`.

**Comment by ferranpujolcamins** [comment]
I find the '|>' and '<|' to be a bit distracting, because they are visually "big".
I suggest to use `~` instead.

**Comment by truizlop** [comment]
Looks good to me. Are you planning to support all combinations for this new operator? I mean, supporting something like:

```swift
func  f(a: Int, b: Int, c: Int) -> Int

f~(1, __, __)
f~(1, __, 3)
f~(__, 2, 3)
```

**Comment by ferranpujolcamins** [comment]
Yeah combinatorial explosion is a problem here. To support up to 6 parameters, we'll need 114 functions.

This has to be written with a script for sure. Maybe it's just not practical?

**Comment by ferranpujolcamins** [comment]
For arities > 6 then you need to concatenate the operator:

```swift
f~(1, __, __)~(1, __, __)
```

Not pretty, but serves as a workaround for the rare case where you have a function with an unsupported arity. At the end we have to draw the line somewhere.

**Comment by truizlop** [comment]
Using a script sounds good. I think we are currently supporting things up to 9 or 10 parameters (check zip implementations in Applicative, for instance). I think both the code and the script should be committed to the repo, in case we need to make any changes in the future. An alternative is to only support a single argument to be applied each time, like:

```swift
func f(a: Int, b: Int, c: Int) -> Int

f~(__, 1, __)~(__, 2) == f~(__, 1, 2)
```

However, the fact that the tuple keeps decreasing makes it hard to follow what's going on.

**Comment by ferranpujolcamins** [comment]
Support for up to 10 parameters means 2026 functions 😅

Applying a single parameter a type seems not to be an improvement over
writing a closure and manually apply the parameters.

**Comment by truizlop** [comment]
Ok, let's go for 6 parameters, I think that's within most reasonable use cases.
